### ISSUE TITLE
ws: describe statement type before sending rows

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#19772]
### Feature request

*Part of epic #19771.*

The in-console SQL editor (https://github.com/MaterializeInc/console/issues/221) wants to know whether the rows its receiving are part of a `SELECT` or a `SUBSCRIBE`. This information is usually available in the command tag in the `CommandComplete` message (e.g., `SELECT 1`, `INSERT 5`), but for `SUBSCRIBE` queries, the console needs to know _ahead_ of the `Rows` responses that it should render the rows live. `CommandComplete` is too late.

I propose that we introduce a new `{"type": "CommandStarting", "payload": "{SELECT|SUBSCRIBE|INSERT}"}` message that gets returned by the WS protocol whenever the connection starts processing a new SQL statement, before it sends any `Rows`, `Row`, or `CommandComplete` messages for that statement.

Alternatives proposed include:

  * Parsing SQL client side. (Hard and fragile, especially since users may submit lots of commands in a batch, like `SELECT 1; SUBSCRIBE ...; SELECT 1.)
  * Sniffing out `mz_diff` and `mz_timestamp` columns in the `Rows` message. Less hard than the above, but still a bit fragile. Can be fooled if a user runs a non-`SUBSCRIBE` query with those column names, e.g. `SELECT 1 as mz_diff, mz_now() as mz_timestamp, ...`.
  * Adding an `is_subscribe: bool` to the `Rows` message. Seems less useful and less general than introducing something like `CommandStarting`, but is the most viable of these alternatives.

cc @mjibson @SangJunBak @aru

**Comment by benesch** [comment]
`has_rows` seems good. `is_subscribe` is funnier because .. well is "subscribe" the feature? Will there be other statements that are not literally `SUBSCRIBE` that will set `is_subscribe: true`? Is there a better name we could use? `is_streaming`? `is_often_long_running`?

Maybe another way of putting it: either `SUBSCRIBE` is the only thing that the console wants to sniff out, in which case `payload == "SUBSCRIBE"` feels right to me, or there is a class of `SUBSCRIBE`-like commands, and we should call the sigil `is_<name-of-subscribe-like-commands>` and not `is_subscribe`.

**Comment by mjibson** [comment]
I agree there could be a better name. I like `is_streaming` but don't care at all.

**Comment by aru** [comment]
hey, just wanted to let you know you might have cc'd the wrong person. I don't have access to the console repository but I assume the plan is to create some awesome visualizations that will not be supported by other traditional SQL clients/protocols (jdbc/odbc). Very interesting project overall.

**Comment by benesch** [comment]
> hey, just wanted to let you know you might have cc'd the wrong person. I don't have access to the console repository but I assume the plan is to create some awesome visualizations that will not be supported by other traditional SQL clients/protocols (jdbc/odbc). Very interesting project overall.

Ha, indeed I did! Thank you for the kind words, and sorry for the ping!
